metals/src/main/scala/scala/meta/internal/metals/Compilers.scala

```diff
@@ -329,14 +345,60 @@ class Compilers(
       path: AbsolutePath,
       interactiveSemanticdbs: Option[InteractiveSemanticdbs]
   ): Option[PresentationCompiler] = {
-    val target = buildTargets
-      .inverseSources(path)
-      .orElse(interactiveSemanticdbs.flatMap(_.getBuildTarget(path)))
-    target match {
-      case None =>
-        if (path.isScalaFilename) Some(ramboCompiler)
-        else None
-      case Some(value) => loadCompiler(value)
+
+    def fromBuildTarget: Option[PresentationCompiler] = {
+      val target = buildTargets
+        .inverseSources(path)
+        .orElse(interactiveSemanticdbs.flatMap(_.getBuildTarget(path)))
+      target match {
+        case None =>
+          if (path.isScalaFilename) Some(ramboCompiler)
+          else None
+        case Some(value) => loadCompiler(value)
+      }
+    }
+
+    if (path.isWorksheet) loadWorksheetCompiler(path).orElse(fromBuildTarget)
+    else fromBuildTarget
+  }
+
+  def loadWorksheetCompiler(
+      path: AbsolutePath
+  ): Option[PresentationCompiler] = {
+    worksheetsCache.get(path)
+  }
+
+  def restartWorksheetPresentationCompiler(
+      path: AbsolutePath,
+      classpath: List[Path]
+  ): Unit = {
+    val created = for {
+      targetId <- buildTargets.inverseSources(path)
+      info <- buildTargets.scalaTarget(targetId)
+      isSupported = ScalaVersions.isSupportedScalaVersion(info.scalaVersion)
+      _ = {
+        if (!isSupported) {
+          scribe.warn(s"unsupported Scala ${info.scalaVersion}")
```

Comment:
This code path didn't exist before, why is that? Would it be possible to somehow move the management of compilers to `Compilers.scala`? There seems to be some overlap, maybe we can generate `BuildTargetIdentifier` for worksheets?

Reply:
Not sure what you mean? It's invoked after evaluating the worksheet in case of changed dependencies, which in essence is a build target change. I am not sure how otherwise to trigger on that particular event. 

Alternatively, we could have worksheets as a different target and update the build target if dependencies change, which will be effectively the same with more indirection. We do need to somehow restart the presentation compiler.

> There seems to be some overlap, maybe we can generate BuildTargetIdentifier for worksheets?

I was thinking about having a separate build target for each worksheet, but since the worksheets can be a part of another build target I decided against. It also seemed like it might cause more work really and I planned to have this for the release tomorrow.

Reply:
Sounds good. I agree with your conclusion to keep this as is